src/viper/trans.ml

```diff
@@ -386,30 +460,42 @@ and stmt ctxt (s : M.exp) : seqn =
             (self ctxt Source.no_region, !!id),
             (!!(AddE(!!(FldAcc (self ctxt (s.at), !!id)),
                      intLitE Source.no_region 1)))));
-          !@(ExhaleS (!@(AndE(!@(MacroCall("$Perm", self ctxt at)),
-                              !@(MacroCall("$Inv", self ctxt at))))));
+          !@(ExhaleS (!@(AndE(!@(MacroCall("$Perm", [self ctxt at])),
+                              !@(MacroCall("$Inv", [self ctxt at]))))));
           !@(SeqnS (
               !@([],
                  [
-                   !@(InhaleS (!@(AndE(!@(MacroCall("$Perm", self ctxt at)),
-                                  !@(AndE(!@(MacroCall("$Inv", self ctxt at)),
+                   !@(InhaleS (!@(AndE(!@(MacroCall("$Perm", [self ctxt at])),
+                                  !@(AndE(!@(MacroCall("$Inv", [self ctxt at])),
                                           !@(GtCmpE(!@(FldAcc (self ctxt at, !@id)),
                                                intLitE Source.no_region 0))))))));
                    !@(FieldAssignS(
                           (self ctxt at, !@id),
                           (!@(SubE(!@(FldAcc (self ctxt at, !@id)),
                                    intLitE at 1)))));
                    !!! (e.at) (SeqnS stmts);
-                   !@(ExhaleS (!@(AndE(!@(MacroCall("$Perm", self ctxt at)),
-                                       !@(MacroCall("$Inv", self ctxt at)))))) ])));
-          !!(InhaleS (!!(AndE(!!(MacroCall("$Perm", self ctxt at)),
-                              !!(MacroCall("$Inv", self ctxt at))))));
+                   !@(ExhaleS (!@(AndE(!@(MacroCall("$Perm", [self ctxt at])),
+                                       !@(MacroCall("$Inv", [self ctxt at])))))) ])));
+          !!(InhaleS (!!(AndE(!!(MacroCall("$Perm", [self ctxt at])),
+                              !!(MacroCall("$Inv", [self ctxt at]))))));
         ])
   | M.WhileE(e, s1) ->
      let (invs, s1') = extract_loop_invariants s1 in
      let invs' = List.map (fun inv -> exp ctxt inv) invs in
      !!([],
         [ !!(WhileS(exp ctxt e, invs', stmt ctxt s1')) ])
+  (* TODO: remove by Motoko->Motoko transformation *)
```

Comment:
`arr := [1, 2, 3]` could be transformed into
```
let t1 := [1, 2, 3]
arr := t1
```
So all array literal will be allocated once and have fresh name.

Reply:
Also note that `arr := [...]` could not produced just by `array_alloc`. We **have to** use temporary variable `t1`.
Otherwise two assignment could produce inconsistent:
```
arr := []
arr := [1]
```
translated (without temporary declaration) to:
```
inhale len(arr) == 0 && array_acc_mut(arr)
exhale array_acc_mut(arr) // we cannot exhale len(arr) == 0
inhale len(arr) == 1 && array_acc_mut(arr)
assert 1 == 0
```
reassignment `arr := t1` solve this problem